src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -318,3 +379,227 @@ ExecuteSelectTasksIntoTupleStore(List *taskList, TupleDesc resultDescriptor)
+
+		ShardPlacement *targetPlacement = CitusMakeNode(ShardPlacement);
+		targetPlacement->nodeName = workerNode->workerName;
+		targetPlacement->nodePort = workerNode->workerPort;
+		targetPlacement->groupId = workerNode->groupId;
+
+		Task *task = CitusMakeNode(Task);
+		task->taskType = SELECT_TASK;
+		task->queryString = QueryStringForFragmentsTransfer(fragmentsTransfer);
+		task->taskPlacementList = list_make1(targetPlacement);
+
+		fetchTaskList = lappend(fetchTaskList, task);
+	}
+
+	return fetchTaskList;
+}
+
+
+/*
+ * QueryStringForFragmentsTransfer returns a query which fetches distributed
+ * result fragments from source node to target node. See the structure of
+ * NodeToNodeFragmentsTransfer for details of how these are decided.
+ */
+static char *
+QueryStringForFragmentsTransfer(NodeToNodeFragmentsTransfer *fragmentsTransfer)
+{
+	ListCell *fragmentCell = NULL;
+	StringInfo queryString = makeStringInfo();
+	StringInfo fragmentNamesArrayString = makeStringInfo();
+	int fragmentCount = 0;
+	NodePair *nodePair = &fragmentsTransfer->nodes;
+	WorkerNode *sourceNode = LookupNodeByNodeId(nodePair->sourceNodeId);
+
+	appendStringInfoString(fragmentNamesArrayString, "ARRAY[");
+
+	foreach(fragmentCell, fragmentsTransfer->fragmentList)
+	{
+		DistributedResultFragment *fragment = lfirst(fragmentCell);
+		char *fragmentName = fragment->resultId;
+
+		if (fragmentCount > 0)
+		{
+			appendStringInfoString(fragmentNamesArrayString, ",");
+		}
+
+		appendStringInfoString(fragmentNamesArrayString,
+							   quote_literal_cstr(fragmentName));
+
+		fragmentCount++;
+	}
+
+	appendStringInfoString(fragmentNamesArrayString, "]::text[]");
+
+	appendStringInfo(queryString,
+					 "SELECT fetch_intermediate_results(%s,%s,%d) bytes",
+					 fragmentNamesArrayString->data,
+					 quote_literal_cstr(sourceNode->workerName),
+					 sourceNode->workerPort);
+
+	ereport(DEBUG3, (errmsg("fetch task on %s:%d: %s", sourceNode->workerName,
+							sourceNode->workerPort, queryString->data)));
+
+	return queryString->data;
+}
+
+
+/*
+ * ExecuteFetchTaskList executes a list of fetch_intermediate_results() tasks
+ * and returns the total byte count.
+ */
+static uint64
+ExecuteFetchTaskList(List *taskList)
+{
+	TupleDesc resultDescriptor = NULL;
+	Tuplestorestate *resultStore = NULL;
+	int resultColumnCount = 1;
+
+#if PG_VERSION_NUM >= 120000
+	resultDescriptor = CreateTemplateTupleDesc(resultColumnCount);
+#else
+	resultDescriptor = CreateTemplateTupleDesc(resultColumnCount, false);
+#endif
+
+	TupleDescInitEntry(resultDescriptor, (AttrNumber) 1, "byte_count", INT8OID, -1, 0);
+
+	resultStore = ExecuteSelectTasksIntoTupleStore(taskList, resultDescriptor);
+
+	TupleTableSlot *slot = MakeSingleTupleTableSlotCompat(resultDescriptor,
+														  &TTSOpsMinimalTuple);
+
+	uint64 byteCount = 0;
+	while (tuplestore_gettupleslot(resultStore, true, false, slot))
+	{
+		bool isNull = false;
+		byteCount += DatumGetInt64(slot_getattr(slot, 1, &isNull));
```

Comment:
Are we ever planning to use the return value of this function?

Reply:
I don't think so. It returns void now.